### PR TITLE
Export LLVM Disassembler

### PR DESCRIPTION
**Pull request by Keno**
https://github.com/Keno/Hooking.jl/issues/2

### CODE REVIEW

**Comment by yuyichao** [comment]
What is `find_region` used for?

**Comment by Keno** [comment]
Checking whether a value that DWARF says is a julia value is actually in the heap.

**Comment by yuyichao** [comment]
FYI, it doesn't handle large object (or `MEMDEBUG`) and shouldn't really be used outside base for anything other than debugging (whether it is good to use it in the debugger is a little borderline...)

**Comment by yuyichao** [comment]
(And it also doesn't cover symbols...)

**Comment by vtjnash** [comment]
i assume you're trying to figure out whether an address is valid? `find_region` can only answer variants on the question "after committing memory at `ptr`, what data-structures need to be updated". i think it would be more accurate to ask the vm if the address can be de-referenced.

**Comment by yuyichao** [comment]
would calling `msync` before each dereferincing be too expensive? (Or maybe use the machinery we already have to catch a segfault without collection backtrace)

**Comment by Keno** [comment]
I already ask the VM if the address can be deferenced, but the problem is that you frequently get valid pointers to invalid julia objects, e.g. when the unwinder didn't correctly restore a register and has a C pointer there instead. I just want a heuristic way to decide if something looks like a valid julia object.

**Comment by yuyichao** [comment]
If you already have save memory derefencing a better way to check if a pointer might be a valid julia object is probably `jl_typeof(jl_typeof(p)) == jl_datatype_type`.

**Comment by Keno** [comment]
> If you already have save memory derefencing a better way to check if a pointer might be a valid julia heap object is probably jl_typeof(jl_typeof(p)) == jl_datatype_type.

Ok, sounds reasonable.

File: src/jlapi.c [hunk 1]
```diff
@@ -5,12 +5,15 @@
...
 */
+#define __STDC_LIMIT_MACROS
+#define __STDC_CONSTANT_MACROS
 #include "platform.h"
 
 #include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
 #include <assert.h>
+#include <llvm-c/Disassembler.h>
```
Comment (vtjnash) [inline src/jlapi.c]: llvm content is not allowed in this file (breaks `make JULIACODEGEN=none` build`

File: src/disasm.cpp [hunk 2]
```diff
@@ -271,7 +271,14 @@ static void print_source_line(raw_ostream &stream, DebugLoc Loc)
 }
 #endif
 
-extern "C"
+extern "C" {
+JL_DLLEXPORT void jl_init_disasm() {
```
Comment (vtjnash) [inline src/disasm.cpp]: are these functions expensive? it would be perhaps be easier to just call this from jl_codegen_init once and not think about again

Comment (Keno) [inline src/disasm.cpp]: They're probably not particularly expensive, they just set up a couple of data structures. Will move then to codegen_init.

**Pull request merged**